.github/workflows/r.yml

```
@@ -52,6 +52,71 @@ env:
   DOCKER_VOLUME_PREFIX: ".docker/"
 
 jobs:
+  ubuntu-cpp-versions:
+    name: Arrow C++ Versions (${{ matrix.config.cpp_version }})
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        config:
+        - { cpp_version: "15.0.0" }
+        - { cpp_version: "14.0.2" }
+        - { cpp_version: "13.0.0" }
+
+    env:
+      ARROW_HOME: "${{ github.workspace }}/arrow"
+
+    steps:
+    - name: Checkout Arrow
+      uses: actions/checkout@3df4ab11eba7bda6032a0b82a6bb43b11571feac # v4.0.0
+      with:
+        path: src
+
+    - name: Cache Arrow C++ Build
+      id: cache-arrow-build
+      uses: actions/cache@v3
+      with:
+        path: arrow
+        key: arrow-r-cpp-release-${{ matrix.config.cpp_version }}-0
+
+    - name: Build Arrow C++ (${{ matrix.config.cpp_version }})
+      if: steps.cache-arrow-build.outputs.cache-hit != 'true'
+      run: |
+        curl -L https://github.com/apache/arrow/archive/refs/tags/apache-arrow-${{ matrix.config.cpp_version }}.tar.gz | \
```

Could you use the official source archive instead of auto generated source archive?
```suggestion
        curl -L "https://www.apache.org/dyn/closer.lua?action=download&filename=arrow/arrow-${{ matrix.config.cpp_version }}/apache-arrow-${{ matrix.config.cpp_version }}.tar.gz" | \
```

---

Or how about using the official Ubuntu binaries?

```bash
sudo apt update
sudo apt install -y -V ca-certificates lsb-release wget
wget https://apache.jfrog.io/artifactory/arrow/$(lsb_release --id --short | tr 'A-Z' 'a-z')/apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb
sudo apt install -y -V ./apache-arrow-apt-source-latest-$(lsb_release --codename --short).deb
sudo apt update
sudo apt install -y -V libarrow-dataset-dev=${{ matrix.config.cpp_version }}-1
```